Backport Trino on Ice blogs (#331)
trinodb/trino.io: _posts/2021-08-12-deep-dive-into-iceberg-internals.md

---
layout: post
title:  "Trino on ice IV: Deep dive into Iceberg internals"
author: Brian Olsen
excerpt_separator: <!--more-->
canonical_url: https://blog.starburstdata.com/trino-on-ice-i-a-gentle-introduction-to-iceberg
---

<p align="center">
 <img align="center" width="100%" height="100%" src="/assets/blog/trino-on-ice/trino-iceberg.png"/>
</p>

Welcome to the Trino on ice series, covering the details around how the Iceberg
table format works with the Trino query engine. The examples build on each

> previous blog, so it's recommended to read the blogs sequentially and reference

Reviewer: the posts

them as needed later. Here are links to the posts in this series:

> * [Trino on ice I: A gentle introduction To Iceberg]({% post_url 2021-05-03-a-gentle-introduction-to-iceberg %})

Reviewer: to Iceberg

how these files work, not only to help understand how Iceberg works, but also to
aid in troubleshooting issues, should you have any issues during ingestion or
querying of your Iceberg table. I like to think of this type of debugging much
like a fun game of operation, and you’re looking to see what causes the red
errors to fly by on your screen.

<!--more-->

![](/assets/blog/trino-on-ice/operation.gif)

## Understanding Iceberg metadata

Iceberg can use any compatible metastore, but for Trino, it only supports the 
Hive metastore and AWS Glue similar to the Hive connector. This is because there
is already a vast amount of testing and support for using the Hive metastore in
Trino. Likewise, many Trino use cases that currently use data lakes already use
the Hive connector and therefore the Hive metastore. This makes it convenient to
have as the leading supported use case as existing users can easily migrate
between Hive to Iceberg tables. Since there is no indication of which connector
is actually executed in the diagram of the Hive connector architecture, it
serves as a diagram that can be used for both Hive and Iceberg. The only

> difference is the plugin used, but if you create a table in Hive, you will also

Reviewer: the connector used

be able to view the same table in Iceberg.

![](/assets/blog/trino-on-ice/iceberg-metadata.png)

To recap the steps taken from the first three blogs; the first blog created an
events table, while the first two blogs ran two insert statements. The first
insert contained three records, while the second insert contained a single
record.

![](/assets/blog/trino-on-ice/iceberg-snapshot-files.png)

Up until this point, the state of the files in MinIO haven’t really been shown
except some of the manifest list pointers from the snapshot in the third blog
post. Using the [MinIO client tool](https://docs.min.io/minio/baremetal/reference/minio-cli/minio-mc.html),

> list the files that Iceberg generated through all these operations and then try

Reviewer: you can list files

to understand what purpose they are serving.

```
% mc tree -f local/
local/
└─ iceberg
   └─ logging.db
      └─ events
         ├─ data
         │  ├─ event_time_day=2021-04-01
         │  │  ├─ 51eb1ea6-266b-490f-8bca-c63391f02d10.orc
         │  │  └─ cbcf052d-240d-4881-8a68-2bbc0f7e5233.orc
         │  └─ event_time_day=2021-04-02
         │     └─ b012ec20-bbdd-47f5-89d3-57b9e32ea9eb.orc
         └─ metadata
            ├─ 00000-c5cfaab4-f82f-4351-b2a5-bd0e241f84bc.metadata.json
            ├─ 00001-27c8c2d1-fdbb-429d-9263-3654d818250e.metadata.json
            ├─ 00002-33d69acc-94cb-44bc-b2a1-71120e749d9a.metadata.json
            ├─ 23cc980c-9570-42ed-85cf-8658fda2727d-m0.avro
            ├─ 92382234-a4a6-4a1b-bc9b-24839472c2f6-m0.avro
            ├─ snap-2720489016575682283-1-92382234-a4a6-4a1b-bc9b-24839472c2f6.avro
            ├─ snap-4564366177504223943-1-23cc980c-9570-42ed-85cf-8658fda2727d.avro
            └─ snap-6967685587675910019-1-bcbe9133-c51c-42a9-9c73-f5b745702cb0.avro
```

> There are a lot of files here, but here are a couple of patterns that you will

Reviewer: that you see

Reviewer: or 

that you can observe:

see with these files.

> First, the top two directories are named data and metadata.

Reviewer: ```suggestion
First, the top two directories are named `data` and `metadata`.
```

`/<bucket>/<database>/<table>/data//<bucket>/<database>/<table>/metadata/`

> As you might expect, data contains the actual ORC files split by partition. This

Reviewer: ```suggestion
As you might expect, `data` contains the actual ORC files split by partition. This
```

is akin to what you would see in a Hive table data directory. What is really of

> interest here is the metadata directory. There are specifically three patterns

Reviewer: ```suggestion
interest here is the `metadata` directory. There are specifically three patterns
```

of files you’ll find here.

`/<bucket>/<database>/<table>/metadata/<file-id>.avro/<bucket>/<database>/<table>/metadata/snap-<snapshot-id>-<version>-<file-id>.avro`

`/<bucket>/<database>/<table>/metadata/<version>-<commit-UUID>.metadata.json`

Iceberg has a persistent tree structure that manages various snapshots of the
data that are created for every mutation of the data. This enables not only a
concurrency model that supports serializable isolation, but also cool features
like time travel across a linear progression of snapshots.

![](/assets/blog/trino-on-ice/iceberg-metastore-files.png)

> This tree structure contains two types of avro files, manifest lists and

Reviewer: isnt it "Avro files"

manifest files. Manifest list files contain pointers to various manifest files

> and the manifest files themselves point to various data files. This blog starts

Reviewer: This post

Reviewer: Also .. in general.. you should not talk about the post as a thing as much

out by covering these manifest files, and later covers the table metadata files
that are suffixed by `.metadata.json`.

[The last blog covered]({% post_url 2021-07-30-iceberg-concurrency-snapshots-spec %})

> the command in Trino that shows the

Reviewer: source wrapping

```

Result:

| snapshots |
| --- |
| s3a://iceberg/logging.db/events/metadata/snap-6967685587675910019-1-bcbe9133-c51c-42a9-9c73-f5b745702cb0.avro |
| s3a://iceberg/logging.db/events/metadata/snap-2720489016575682283-1-92382234-a4a6-4a1b-bc9b-24839472c2f6.avro |
| s3a://iceberg/logging.db/events/metadata/snap-4564366177504223943-1-23cc980c-9570-42ed-85cf-8658fda2727d.avro |

You’ll notice that the manifest list returns the Avro files prefixed with
`snap-` are returned. These files are directly correlated with the snapshot
record stored in the metastore. According to the diagram above, snapshots are
records in the metastore that contain the url of the manifest list in the Avro
file. Avro files are binary files and not something you can just open up in a
text editor to read. Using the 
[avro-tools.jar tool](https://downloads.apache.org/avro/avro-1.10.2/java/avro-tools-1.10.2.jar)
distributed by the 
[Apache Avro project](https://avro.apache.org/docs/current/index.html),
you can actually inspect the contents of this file to get a better understanding
of how it is used by Iceberg.

The first snapshot is generated on the creation of the events table. Upon

> inspecting this file, you will notice that the file is empty. The output is an

Reviewer: you notice that

> empty line that the jq json command line utility removes on pretty printing the

Reviewer: ```suggestion
empty line that the `jq` JSON command line utility removes on pretty printing the
```

Reviewer: also do a general replace .. it should always be JSON

json that is returned, which is just a newline. This snapshot represents an

> empty state of the table upon creation. In order to investigate these snapshots,

Reviewer: Cut the clutter

To investigate the snapshots you need to download the files to your local filesystem. Let's move them to the home directory:

it should also be noted that you will need to download these files on your local
filesystem. In this blog we’ll just move them to the home directory `~/`.

```
% java -jar  ~/Desktop/avro_files/avro-tools-1.10.0.jar tojson ~/snap-6967685587675910019-1-bcbe9133-c51c-42a9-9c73-f5b745702cb0.avro | jq .
````

Result:

```

> 

Reviewer: ?? something missing here?

Reviewer: at minimum we need to mention that the result is empty

Author: I mention this above

contents of a manifest list.

```
% java -jar  ~/Desktop/avro_files/avro-tools-1.10.0.jar tojson ~/snap-2720489016575682283-1-92382234-a4a6-4a1b-bc9b-24839472c2f6.avro | jq .
```

Result:

```
{
   "manifest_path":"s3a://iceberg/logging.db/events/metadata/92382234-a4a6-4a1b-bc9b-24839472c2f6-m0.avro",
   "manifest_length":6114,
   "partition_spec_id":0,
   "added_snapshot_id":{
      "long":2720489016575682000
   },
   "added_data_files_count":{
      "int":2
   },
   "existing_data_files_count":{
      "int":0
   },
   "deleted_data_files_count":{
      "int":0
   },
   "partitions":{
      "array":[
         {
            "contains_null":false,
            "lower_bound":{
               "bytes":"\u001eI\u0000\u0000"
            },
            "upper_bound":{
               "bytes":"\u001fI\u0000\u0000"
            }
         }
      ]
   },
   "added_rows_count":{
      "long":3
   },
   "existing_rows_count":{
      "long":0
   },
   "deleted_rows_count":{
      "long":0
   }
}
```

To understand each of the values in each of these rows, you can refer to the 
Iceberg 
[specification in the manifest list file section](https://iceberg.apache.org/spec/#manifest-lists).

> Instead of covering these exhaustively, the focus will be on a few key fields.

Reviewer: , let's focus on

```

Result: 

```
{
   "manifest_path":"s3a://iceberg/logging.db/events/metadata/23cc980c-9570-42ed-85cf-8658fda2727d-m0.avro",
   "partition_spec_id":0,
   "added_snapshot_id":{
      "long":4564366177504223700
   },
   "partitions":{
      "array":[
         {
            "contains_null":false,
            "lower_bound":{
               "bytes":"\u001eI\u0000\u0000"
            },
            "upper_bound":{
               "bytes":"\u001eI\u0000\u0000"
            }
         }
      ]
   },
   "added_rows_count":{
      "long":1
   }
}
{
   "manifest_path":"s3a://iceberg/logging.db/events/metadata/92382234-a4a6-4a1b-bc9b-24839472c2f6-m0.avro",
   "partition_spec_id":0,
   "added_snapshot_id":{
      "long":2720489016575682000
   },
   "partitions":{
      "array":[
         {
            "contains_null":false,
            "lower_bound":{
               "bytes":"\u001eI\u0000\u0000"
            },
            "upper_bound":{
               "bytes":"\u001fI\u0000\u0000"
            }
         }
      ]
   },
   "added_rows_count":{
      "long":3
   }
}
```

> In the listing of the manifest file related to the last snapshot, you will

Reviewer: Continues to have lots of "will"

            }
         ]
      },
      "null_value_counts":{
         "array":[
            {
               "key":1,
               "value":0
            },
            {
               "key":2,
               "value":0
            },
            {
               "key":3,
               "value":0
            },
            {
               "key":4,
               "value":0
            }
         ]
      },
      "nan_value_counts":null,
      "lower_bounds":{
         "array":[
            {
               "key":1,
               "value":"ERROR"
            },
            {
               "key":3,
               "value":"Double oh noes"
            }
         ]
      },
      "upper_bounds":{
         "array":[
            {
               "key":1,
               "value":"WARN"
            },
            {
               "key":3,
               "value":"Maybeh oh noes?"
            }
         ]
      },
      "key_metadata":null,
      "split_offsets":null
   }
}
```

Now this is a very big output, but in summary, there’s really not too much to
these files. As before, there is a 
[Manifest section in the Iceberg spec](https://iceberg.apache.org/spec/#manifests)

> that will detail what each of these fields means. Here are the important fields in the context of this blog post.

Reviewer: ```suggestion
that details what each of these fields means. Here are the important fields:
```

* `data_file.*_count` - Multiple fields that contain a map from column id to 
number of values, null, nan counts in the file. These can be used to quickly 
filter out unnecessary get operations.
* `data_file.*_bounds` - Multiple fields that contain a map from column id to
lower or upper bound in the column serialized as binary. Each value must be less
than or equal to all non-null, non-NaN values in the column for the file.

Each data file struct contains a partition and data file that it maps to. These
files will only be scanned and returned if the criteria for the query is met
when checking all of the count, bounds, and other statistics that are recorded
in the file. Ideally only files that contain data relevant to the query should
be scanned at all. Having information like the record count may also help in the
query planning process to determine splits and other information. This
particular optimization hasn’t been completed yet as planning typically happens
before traversal of the files. It is still in ongoing discussion and
[is discussed a bit by Iceberg creator Ryan Blue in a recent meetup](https://youtu.be/ifXpOn0NJWk?t=2132)

> he joined at a Trino Meetup. If this is something you are interested in, keep

Reviewer: remove 

`he joined at a Trino Meetup`

posted on the Slack channel and releases as the Trino Iceberg connector
progresses in this area.

As mentioned above, the last set of files that you find in the metadata

> directory which are suffixed with .metadata.json. These files at baseline are a

Reviewer: ```suggestion
directory which are suffixed with `.metadata.json`. These files at baseline are a
```

|--------------------------|-------------------------------------------------------------------------------------------------|
|EXTERNAL                  |TRUE                                                                                             |
|metadata_location         |s3a://iceberg/logging.db/events/metadata/00002-33d69acc-94cb-44bc-b2a1-71120e749d9a.metadata.json|
|numFiles                  |2                                                                                                |
|previous_metadata_location|s3a://iceberg/logging.db/events/metadata/00001-27c8c2d1-fdbb-429d-9263-3654d818250e.metadata.json|
|table_type                |iceberg                                                                                          |
|totalSize                 |5323                                                                                             |
|transient_lastDdlTime     |1622865672                                                                                       |

So as you can see, the metastore is saying the current metadata location is the
`00002-33d69acc-94cb-44bc-b2a1-71120e749d9a.metadata.json` file. Now you can

> dive in to see the table metadata that is currently being used by the Iceberg

Reviewer: drop "currently"

            "added-records":"1",
            "added-files-size":"746",
            "changed-partition-count":"1",
            "total-records":"4",
            "total-data-files":"3",
            "total-delete-files":"0",
            "total-position-deletes":"0",
            "total-equality-deletes":"0"
         },
         "manifest-list":"s3a://iceberg/logging.db/events/metadata/snap-4564366177504223943-1-23cc980c-9570-42ed-85cf-8658fda2727d.avro"
      }
   ],
   "snapshot-log":[
      {
         "timestamp-ms":1622865672882,
         "snapshot-id":6967685587675910019
      },
      {
         "timestamp-ms":1622865680419,
         "snapshot-id":2720489016575682283
      },
      {
         "timestamp-ms":1622865686278,
         "snapshot-id":4564366177504223943
      }
   ],
   "metadata-log":[
      {
         "timestamp-ms":1622865672894,
         "metadata-file":"s3a://iceberg/logging.db/events/metadata/00000-c5cfaab4-f82f-4351-b2a5-bd0e241f84bc.metadata.json"
      },
      {
         "timestamp-ms":1622865680524,
         "metadata-file":"s3a://iceberg/logging.db/events/metadata/00001-27c8c2d1-fdbb-429d-9263-3654d818250e.metadata.json"
      }
   ]
}
```

> As you can see, these json files can quickly grow as you perform different

Reviewer: JSON

updates on your table. This file contains a pointer to all of the snapshots and
manifest list files, much like the output you found from looking at the
snapshots in the table. A really important piece to note is the schema is stored
here. This is what Trino uses for validation on inserts and reads. As you may
expect, there is the root location of the table itself, as well as a unique
table identifier. The final part I’d like to note about this file is the
partition-spec and partition-specs fields. The partition-spec field holds the
current partition spec, while the partition-specs is an array that can hold a
list of all partition specs that have existed for this table. As pointed out
earlier, you can have many different manifest files that use different partition
specs. That wraps up all of the metadata file types you can expect to see in
Iceberg!

This post wraps up the Trino on Ice series. Hopefully these blog posts serve as
a helpful initial dialogue about what is expected to grow as a vital portion of

> an open data lake stack. What are you waiting for? Come join the fun and help us

Reviewer: open data lakehouse stack
